RFC-lifecycle.md

```diff
@@ -0,0 +1,65 @@
+| ----------------------------------------------------------------------- | ---------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
+| Assert that all mocks have been satisfied                               | `scopes.forEach(scope => scope.done())`                          | Meh. This requires keeping track of all the scopes.                                                                            |
+| Completely reset `nock` after a test                                    | `nock.restore(); nock.cleanAll(); nock.enableNetConnect()`       | Meh. Too much typing.                                                                                                          |
+| Forbid unmocked requests                                                | `nock.disableNetConnect()`                                       | Meh. Errors are swallowed up (#884). Also, if `restore()` has been called does `disableNetConnect()` automatically reactivate? |
+| Allow unmocked requests, but only to certain hosts                      | `nock.disableNetConnect(); nock.enableNetConnect('example.com')` | Should this require two calls?                                                                                                 |
+| Simulate network connection failure                                     | N/A                                                              | Meh. This is possible with `disableNetConnect()` however it's muddied up between this and "Forbid unmocked requests" (#884).   |
+| Temporarily disable http interception while preserving registered mocks | `nock.restore()`                                                 | This is a confusing name, as it only cleans _part_ of nock's state.                                                            |
+
+The code for these use cases should be further reviewed and validated. I'm not
+completely sure it's correct.
+
+# Proposed solution
+
+Create functions that correspond to these use cases, and give them unambiguous names.
+
+1. Add `nock.reset()` which resets 100% of nock's state. I think that is the
+   equivalent of `nock.restore(); nock.cleanAll(); nock.enableNetConnect()`.
+   This is suitable for running from `afterEach()` or `finally`.
+2. Rename `nock.restore()` to `nock.deactivate()`. `nock.restore()` should
+   still work but emit a deprecation warning to encourage switching to
+   `nock.reset()`. The new name harmonizes with `nock.activate()` as they are
+   inverses of each other.
+3. Leave `nock.cleanAll()` as is.
+4. Add `nock.assertAll()` which does the equivalent of
+   `scopes.forEach(scope => scope.done())`. This probably should be invoked
+   in the test itself.
+5. Rename `scope.done()` to `scope.assert()` to harmonize with `assertAll()`
```

Comment:
I’m not sure about the naming of `scope.assert()`, it reads as if I should pass in something. Maybe `scope.assertDone()` and `scope.assertAllDone()`?

Reply:
Hmm yea it does seem like it wants an argument. `nock.assertAllDone()` and `scope.assertDone()` could work. It's a little long but maybe that's fine.

For someone encountering the test code for the first time, it's definitely clearer for than `done()`.

Reply:
I’m a fan of clear method names, even if they are long. Code readability is king 👑

Reply:
Are there any other mock libraries with interfaces you like? I'm curious to compare. I looked at Sinon, but it doesn't really have assertion methods like this.

"Done" isn't the clearest either. Really what we want to say is "expectations fulfilled" or "requests made" or "mocks consumed." I'm 100% agreed that clear long names are better than short names, though I don't want the names to be longer than necessary.

Reply:
Another option is `nock.verifyAll()` / `scope.verify()`.

I don’t think it has the same problem as `assert` of wanting an argument.

Or: `nock.verifyAllRequests()` / `scope.verifyRequest()`

Reply:
I'll keep thinking about it too.

Reply:
how about expressing that all interceptors have been used up, the naming would align with the READ THIS! section in the readme: https://github.com/nock/nock#read-this---about-interceptors

sth like `nock.allInterceptorsUsed()`? 

I wonder if that method is necessary at all, maybe we should reduce the amount of APIs. 

Maybe something like the `.remainingInterceptors()` method would be enough, then I do the assertion myself in my tests?

Reply:
`assertAllInterceptorsUsed()` seems good. Though is it scopes or interceptors we're talking about? Maybe `assertAllMocksUsed()` is better.

We do have `nock.pendingMocks()`, so it's possible if people want control of the assertion they can add it.

Though I would rather not write my own assertion here. It ends up being boilerplate code that has to be replicated across each library. Nock is in a good position to write helpful error messages. Down the line you could imagine saying something more detailed like, "GET http://example.test:80/ expected two calls but only got one", which a caller wouldn't be able to do. I don't like adding unnecessary APIs, though I think nock is the place where this assert function should be implemented.

Reply:
I changed `nock.assertAll()` to `nock.assertAllMocksUsed()` and `scope.assert()` to `scope.assertMocksUsed()`.

Reply:
now seeing that, I wonder why one of the method is called `assertAllMocksUsed` and the other `assertMocksUsed`. Wouldn’t it be more consistent to call both `assertMocksUsed` and the "scope" of the method call would be clear based on wether you call it on a scope instance or the global `nock` object?

Reply:
It’s a good question. If the names are the same I think it’s too likely someone will call the wrong one. Since `scope.done()` was the way to do this for a long time, I don’t want to provide two different ways which are as confusable as `scope.assertMocksUsed()` and `nock.assertMocksUsed()`. Especially if those two functions are doing different things.

Reply:
Another option: `scope.assertUsed()`, `nock.assertAllMocksUsed()`.

Reply:
> Especially if those two functions are doing different things

Wouldn’t both functions do the same thing? `nock.assertMocksUsed()` would check globally, `scope.assertMocksUsed()` would check for only that scope?